File: customfit-reactnative-client-sdk/src/core/model/CFUser.ts
import { 
  CFUser, 
  CFUserBuilder, 
  EvaluationContext, 
  DeviceContext, 
  ApplicationInfo, 
  ContextType,
  evaluationContextToMap,
  deviceContextToMap,
  applicationInfoToMap
} from '../types/CFTypes';

/**
 * User model implementation for CustomFit SDK
 */
export class CFUserImpl implements CFUser {
  userCustomerId?: string;
  anonymousId?: string;
  deviceId?: string;
  anonymous?: boolean;
  properties?: Record<string, any>;
  contexts?: EvaluationContext[];
  device?: DeviceContext;
  application?: ApplicationInfo;

  constructor(
    userCustomerId?: string,
    anonymousId?: string,
    deviceId?: string,
    anonymous?: boolean,
    properties?: Record<string, any>,
    contexts?: EvaluationContext[],
    device?: DeviceContext,
    application?: ApplicationInfo
  ) {
    this.userCustomerId = userCustomerId;
    this.anonymousId = anonymousId;
    this.deviceId = deviceId;
    this.anonymous = anonymous ?? false;
    this.properties = properties ?? {};
    this.contexts = contexts ?? [];
    this.device = device;
    this.application = application;
  }

  /**
   * Create a new user with updated user customer ID
   */
  withUserCustomerId(userCustomerId: string): CFUser {
    return new CFUserImpl(
      userCustomerId,
      this.anonymousId,
      this.deviceId,
      this.anonymous,
      this.properties,
      this.contexts,
      this.device,
      this.application
    );
  }

  /**
   * Create a new user with updated anonymous ID
   */
  withAnonymousId(anonymousId: string): CFUser {
    return new CFUserImpl(
      this.userCustomerId,
      anonymousId,
      this.deviceId,
      this.anonymous,
      this.properties,
      this.contexts,
      this.device,
      this.application
    );
  }

  /**
   * Create a new user with updated device ID
   */
  withDeviceId(deviceId: string): CFUser {
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      deviceId,
      this.anonymous,
      this.properties,
      this.contexts,
      this.device,
      this.application
    );
  }

  /**
   * Create a new user with updated anonymous status
   */
  withAnonymous(anonymous: boolean): CFUser {
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      this.deviceId,
      anonymous,
      this.properties,
      this.contexts,
      this.device,
      this.application
    );
  }

  /**
   * Create a new user with updated properties
   */
  withProperties(properties: Record<string, any>): CFUser {
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      this.deviceId,
      this.anonymous,
      { ...this.properties, ...properties },
      this.contexts,
      this.device,
      this.application
    );
  }

  /**
   * Create a new user with a single updated property
   */
  withProperty(key: string, value: any): CFUser {
    return this.withProperties({ [key]: value });
  }

  /**
   * Create a new user with an added context
   */
  withContext(context: EvaluationContext): CFUser {
    const updatedContexts = [...(this.contexts || []), context];
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      this.deviceId,
      this.anonymous,
      this.properties,
      updatedContexts,
      this.device,
      this.application
    );
  }

  /**
   * Create a new user with updated device context
   */
  withDeviceContext(device: DeviceContext): CFUser {
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      this.deviceId,
      this.anonymous,
      this.properties,
      this.contexts,
      device,
      this.application
    );
  }

  /**
   * Create a new user with updated application info
   */
  withApplicationInfo(application: ApplicationInfo): CFUser {
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      this.deviceId,
      this.anonymous,
      this.properties,
      this.contexts,
      this.device,
      application
    );
  }

  /**
   * Create a new user with a context removed
   */
  removeContext(type: ContextType, key: string): CFUser {
    const updatedContexts = (this.contexts || []).filter(
      context => !(context.type === type && context.key === key)
    );
    return new CFUserImpl(
      this.userCustomerId,
      this.anonymousId,
      this.deviceId,
      this.anonymous,
      this.properties,
      updatedContexts,
      this.device,
      this.application
    );
  }

  /**
   * Convert user to a plain object for API calls
   */
  toUserMap(): Record<string, any> {
    // Start with a copy of properties
    const updatedProperties = { ...this.properties };

    // Inject contexts, device, application into properties (if present)
    if (this.contexts && this.contexts.length > 0) {
      updatedProperties.contexts = this.contexts.map(evaluationContextToMap);
    }
    if (this.device) {
      updatedProperties.device = deviceContextToMap(this.device);
    }
    if (this.application) {
      updatedProperties.application = applicationInfoToMap(this.application);
    }

    // Add basic device info if no device context is provided
    if (!this.device) {
      updatedProperties.device = {
        device_id: this.deviceId,
        os_name: 'React Native',
        sdk_type: 'react-native',
        sdk_version: '1.0.0',
      };
    }

    const map: Record<string, any> = {
      anonymous: this.anonymous ?? false,
      properties: updatedProperties,
    };

    if (this.userCustomerId) {
      map.user_customer_id = this.userCustomerId;
    }

    if (this.anonymousId) {
      map.anonymous_id = this.anonymousId;
    }

    return map;
  }

  /**
   * Create a default user instance
   */
  static defaultUser(): CFUser {
    return new CFUserImpl();
  }

  /**
   * Create a builder for constructing users
   */
  static builder(userCustomerId?: string): CFUserBuilder {
    return new CFUserBuilderImpl(userCustomerId);
  }
}

/**
 * Builder for creating CFUser instances
 */
export class CFUserBuilderImpl implements CFUserBuilder {
  private _userCustomerId?: string;
  private _anonymousId?: string;
  private _deviceId?: string;
  private _anonymous?: boolean;
  private _properties?: Record<string, any>;
  private _contexts?: EvaluationContext[];
  private _device?: DeviceContext;
  private _application?: ApplicationInfo;

  constructor(userCustomerId?: string) {
    this._userCustomerId = userCustomerId;
    this._properties = {};
    this._contexts = [];
  }

  userCustomerId(id: string): CFUserBuilder {
    this._userCustomerId = id;
    return this;
  }

  anonymousId(id: string): CFUserBuilder {
    this._anonymousId = id;
    return this;
  }

  deviceId(id: string): CFUserBuilder {
    this._deviceId = id;
    return this;
  }

  anonymous(isAnonymous: boolean): CFUserBuilder {
    this._anonymous = isAnonymous;
    return this;
  }

  properties(props: Record<string, any>): CFUserBuilder {
    this._properties = { ...this._properties, ...props };
    return this;
  }

  property(key: string, value: any): CFUserBuilder {
    if (!this._properties) {
      this._properties = {};
    }
    this._properties[key] = value;
    return this;
  }

  context(context: EvaluationContext): CFUserBuilder {
    if (!this._contexts) {
      this._contexts = [];
    }
    this._contexts.push(context);
    return this;
  }

  deviceContext(device: DeviceContext): CFUserBuilder {
    this._device = device;
    return this;
  }

  applicationInfo(application: ApplicationInfo): CFUserBuilder {
    this._application = application;
    return this;
  }

  build(): CFUser {
    return new CFUserImpl(
      this._userCustomerId,
      this._anonymousId,
      this._deviceId,
      this._anonymous,
      this._properties,
      this._contexts,
      this._device,
      this._application
    );
  }
} 

File: customfit-flutter-client-sdk/lib/src/core/model/cf_user.dart
import 'device_context.dart';
import 'application_info.dart';
import 'evaluation_context.dart';
import 'context_type.dart';
import 'private_attributes_request.dart';

/// User model defining identity and attributes for the CustomFit SDK
class CFUser {
  /// User customer ID
  final String? userCustomerId;

  /// Whether the user is anonymous
  final bool anonymous;

  /// Private fields that should not be sent to analytics
  final PrivateAttributesRequest? privateFields;

  /// Session fields that should not be sent to analytics
  final PrivateAttributesRequest? sessionFields;

  /// User properties as key-value pairs
  final Map<String, dynamic> properties;

  /// Evaluation contexts for the user
  final List<EvaluationContext> contexts;

  /// Device context information
  final DeviceContext? device;

  /// Application information
  final ApplicationInfo? application;

  /// Constructor
  CFUser({
    this.userCustomerId,
    this.anonymous = false,
    this.privateFields,
    this.sessionFields,
    this.properties = const {},
    this.contexts = const [],
    this.device,
    this.application,
  });

  /// Creates a CFUser from a map representation
  factory CFUser.fromMap(Map<String, dynamic> map) {
    return CFUser(
      userCustomerId: map['user_customer_id'] as String?,
      anonymous: map['anonymous'] as bool? ?? false,
      privateFields: map['private_fields'] != null
          ? PrivateAttributesRequest.fromMap(
              map['private_fields'] as Map<String, dynamic>)
          : null,
      sessionFields: map['session_fields'] != null
          ? PrivateAttributesRequest.fromMap(
              map['session_fields'] as Map<String, dynamic>)
          : null,
      properties: (map['properties'] as Map<String, dynamic>?) ?? {},
      contexts: (map['contexts'] as List<dynamic>?)
              ?.map((e) => EvaluationContext.fromMap(e as Map<String, dynamic>))
              .toList() ??
          [],
      device: map['device'] != null
          ? DeviceContext.fromMap(map['device'] as Map<String, dynamic>)
          : null,
      application: map['application'] != null
          ? ApplicationInfo.fromMap(map['application'] as Map<String, dynamic>)
          : null,
    );
  }

  /// Convert user to a map for serialization
  Map<String, dynamic> toMap() {
    // Start with a copy of properties
    final updatedProperties = Map<String, dynamic>.from(properties);

    // Inject contexts, device, application into properties (if present)
    if (contexts.isNotEmpty) {
      updatedProperties['contexts'] = contexts.map((e) => e.toMap()).toList();
    }
    if (device != null) {
      updatedProperties['device'] = device!.toMap();
    }
    if (application != null) {
      updatedProperties['application'] = application!.toMap();
    }

    return {
      'user_customer_id': userCustomerId,
      'anonymous': anonymous,
      if (privateFields != null) 'private_fields': privateFields!.toMap(),
      if (sessionFields != null) 'session_fields': sessionFields!.toMap(),
      'properties': updatedProperties,
    }..removeWhere((_, v) => v == null);
  }

  /// Create a copy with an added property
  CFUser addProperty(String key, dynamic value) {
    final updatedProperties = Map<String, dynamic>.from(properties);
    updatedProperties[key] = value;
    return CFUser(
      userCustomerId: userCustomerId,
      anonymous: anonymous,
      privateFields: privateFields,
      sessionFields: sessionFields,
      properties: updatedProperties,
      contexts: contexts,
      device: device,
      application: application,
    );
  }

  /// Create a copy with added context
  CFUser addContext(EvaluationContext context) {
    final updatedContexts = List<EvaluationContext>.from(contexts);
    updatedContexts.add(context);
    return CFUser(
      userCustomerId: userCustomerId,
      anonymous: anonymous,
      privateFields: privateFields,
      sessionFields: sessionFields,
      properties: properties,
      contexts: updatedContexts,
      device: device,
      application: application,
    );
  }

  /// Create a copy with removed context
  CFUser removeContext(ContextType type, String key) {
    final updatedContexts = contexts
        .where((context) => !(context.type == type && context.key == key))
        .toList();
    return CFUser(
      userCustomerId: userCustomerId,
      anonymous: anonymous,
      privateFields: privateFields,
      sessionFields: sessionFields,
      properties: properties,
      contexts: updatedContexts,
      device: device,
      application: application,
    );
  }

  /// Create a copy with updated device context
  CFUser withDeviceContext(DeviceContext device) {
    return CFUser(
      userCustomerId: userCustomerId,
      anonymous: anonymous,
      privateFields: privateFields,
      sessionFields: sessionFields,
      properties: properties,
      contexts: contexts,
      device: device,
      application: application,
    );
  }

  /// Create a copy with updated application info
  CFUser withApplicationInfo(ApplicationInfo application) {
    return CFUser(
      userCustomerId: userCustomerId,
      anonymous: anonymous,
      privateFields: privateFields,
      sessionFields: sessionFields,
      properties: properties,
      contexts: contexts,
      device: device,
      application: application,
    );
  }
}
